packages/go/dawgs/graph/node.go

```diff
@@ -251,6 +252,199 @@ func (s *NodeSet) UnmarshalJSON(input []byte) error {
 
+type ThreadSafeNodeSet struct {
+	nodeSet NodeSet
+	rwLock  *sync.RWMutex
+}
+
+func NewThreadSafeNodeSet() *ThreadSafeNodeSet {
+	return &ThreadSafeNodeSet{
+		nodeSet: NodeSet{},
+		rwLock:  &sync.RWMutex{},
+	}
+}
+
+// Pick returns a single Node instance from this set. Repeated calls to this function are not guaranteed to return
+// the same Node instance.
+func (s ThreadSafeNodeSet) Pick() *Node {
+	s.rwLock.RLock()
+	defer s.rwLock.RUnlock()
+
+	for _, value := range s.nodeSet {
+		return value
+	}
+
+	return nil
+}
+
+// ByKind returns a new ThreadSafeNodeSet containing only Node instances that contain any one of the given Kind instances.
+func (s ThreadSafeNodeSet) ByKind(kinds ...Kind) NodeSet {
+	s.rwLock.RLock()
+	defer s.rwLock.RUnlock()
+
+	newThreadSafeNodeSet := NewThreadSafeNodeSet()
+
+	for _, node := range s.nodeSet {
+		if node.Kinds.ContainsOneOf(kinds...) {
+			newThreadSafeNodeSet.Add(node)
+		}
+	}
+
+	return newThreadSafeNodeSet.nodeSet
+}
+
+// Remove removes a Node from this set by its database ID.
+func (s ThreadSafeNodeSet) Remove(id ID) {
+	s.rwLock.RLock()
+	defer s.rwLock.RUnlock()
+
+	delete(s.nodeSet, id)
+}
+
+// Get returns a Node from this set by its database ID.
+func (s ThreadSafeNodeSet) Get(id ID) *Node {
+	s.rwLock.RLock()
+	defer s.rwLock.RUnlock()
+
+	return s.nodeSet[id]
+}
+
+// Len returns the number of unique Node instances in this set.
+func (s ThreadSafeNodeSet) Len() int {
+	s.rwLock.RLock()
+	defer s.rwLock.RUnlock()
+
+	return len(s.nodeSet)
+}
+
+// Copy returns a shallow copy of this set.
+func (s ThreadSafeNodeSet) Copy() ThreadSafeNodeSet {
+	s.rwLock.RLock()
+	defer s.rwLock.RUnlock()
+
+	var (
+		newThreadSafeNodeSet = NewThreadSafeNodeSet()
+		newNodeSet           = make(NodeSet, len(s.nodeSet))
+	)
+
+	for k, v := range s.nodeSet {
+		newNodeSet[k] = v
+	}
+
+	newThreadSafeNodeSet.nodeSet = newNodeSet
+
+	return *NewThreadSafeNodeSet()
+}
+
+// KindSet returns a NodeKindSet constructed from the Node instances in this set.
+func (s ThreadSafeNodeSet) KindSet() NodeKindSet {
+	s.rwLock.RLock()
+	defer s.rwLock.RUnlock()
+
+	nodeKindSet := NodeKindSet{}
+
+	for _, node := range s.nodeSet {
+		nodeKindSet.Add(node)
+	}
+
+	return nodeKindSet
+}
+
+// Contains returns true if the ID of the given Node is stored within this NodeSet.
+func (s ThreadSafeNodeSet) Contains(node *Node) bool {
+	s.rwLock.RLock()
+	defer s.rwLock.RUnlock()
+
+	_, seen := s.nodeSet[node.ID]
+
+	return seen
+}
+
+// ContainsID returns true if the Node represented by the given ID is stored within this NodeSet.
+func (s ThreadSafeNodeSet) ContainsID(id ID) bool {
+	s.rwLock.RLock()
+	defer s.rwLock.RUnlock()
+
+	_, seen := s.nodeSet[id]
+
+	return seen
+}
+
+// Add adds a given Node to the NodeSet.
+func (s ThreadSafeNodeSet) Add(nodes ...*Node) {
+	s.rwLock.RLock()
+	defer s.rwLock.RUnlock()
+
+	for _, node := range nodes {
+		s.nodeSet[node.ID] = node
+	}
+}
+
+func (s ThreadSafeNodeSet) CheckedAdd(node *Node) bool {
```

Comment:
All of these methods are mirrors of the non thread safe counterpart, `NodeSet`, so I wanted to keep the method names the same between the two. What do you think? This might also touch on the ByKind(s) comment above.

Reply:
If this a threadsafe version of `NodeSet` I think we should keep the APIs in sync. If you have the bandwidth it might be good to create an interface to ensure that's happening across both types if we're keeping both around.